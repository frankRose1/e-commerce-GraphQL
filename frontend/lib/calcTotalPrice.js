/**
 * Take in a users cart and return the total cost of all items in cart
 * @param {array} cart - array of cartItems which contain data about a specific item in the DB
 * @return {integer} totalPrice - total of cart items
 */
function calcTotalPrice(cart){
  return cart.reduce((tally, cartItem) => {
    //in the case an item is deleted from the DB the cartItem will have an ID that points to a deleted item
    if (!cartItem.item) return tally;
    return tally + (cartItem.quantity * cartItem.item.price);
  }, 0);
}

export default calcTotalPrice;