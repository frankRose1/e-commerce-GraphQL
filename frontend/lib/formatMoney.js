/**
 * Format dollar amount of an item
 * @param {integer} amount - dollar amount of an item in cents
 * @return newly formatted price as a string, US currency 
 */
export default function(amount){
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }
  //for whole dollar amounts, leave off the .00
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const formatter = new Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}