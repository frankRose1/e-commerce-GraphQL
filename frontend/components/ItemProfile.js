import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Link from 'next/link';
import PropTypes from 'prop-types';
import Head from 'next/head';
import AddToCart from './AddToCart';
import DeleteItem from './DeleteItem';
import Error from './ErrorMessage';
import Loading from './Loading';
import formatMoney from '../lib/formatMoney';
import ItemProfileStyles from './styles/ItemProfileStyles';

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      largeImage
      title
      description
      price
    }
  }
`;

class ItemProfile extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired
  };

  render() {
    return (
      <Query
        query={SINGLE_ITEM_QUERY}
        variables={{
          id: this.props.id
        }}
      >
        {({ data, error, loading }) => {
          const { item } = data;
          if (loading) return <Loading />;
          if (error) return <Error error={error} />;
          if (!item) return <p>No item found!</p>;

          return (
            <ItemProfileStyles>
              <Head>
                <title>BackPackzz | {item.title}</title>
              </Head>
              <h3 className='price'>{formatMoney(item.price)}</h3>
              <img src={item.largeImage} alt={item.title} />
              <div className='details'>
                <h2>{item.title}</h2>
                <p>{item.description}</p>

                <div className='item-buttons'>
                  <Link
                    href={{
                      pathname: '/update',
                      query: { id: this.props.id }
                    }}
                  >
                    <a>Edit</a>
                  </Link>
                  <AddToCart itemId={this.props.id} />
                  <DeleteItem id={this.props.id} />
                </div>
              </div>
            </ItemProfileStyles>
          );
        }}
      </Query>
    );
  }
}

export default ItemProfile;
export { SINGLE_ITEM_QUERY };
