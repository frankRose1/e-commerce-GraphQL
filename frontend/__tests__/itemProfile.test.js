import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import ItemProfile, { SINGLE_ITEM_QUERY } from '../components/ItemProfile';
import { fakeItem } from '../lib/testUtils';
import formatMoney from '../lib/formatMoney';

const item = fakeItem();

const mocks = [
  {
    //when a request is made with this query/variable
    request: { query: SINGLE_ITEM_QUERY, variables: { id: 'abc123' } },
    //return this mock data
    result: {
      data: {
        item: item
      }
    }
  }
];

const notFoundMocks = [
  {
    request: { query: SINGLE_ITEM_QUERY, variables: { id: 'noItem' } },
    result: {
      data: {
        item: null
      }
    }
  }
];

describe('<ItemProfile />', () => {
  it('renders with correct data', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ItemProfile id={'abc123'} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('.price').text()).toEqual(formatMoney(item.price));
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
    expect(wrapper.find('AddToCart').exists()).toBe(true);
    expect(wrapper.find('DeleteItem').exists()).toBe(true);
  });

  it('shows an error message if an item isnt found', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notFoundMocks}>
        <ItemProfile id={'noItem'} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const message = wrapper.find('[data-test="graphql-error"]');
    expect(toJSON(message)).toMatchSnapshot();
  });
});
