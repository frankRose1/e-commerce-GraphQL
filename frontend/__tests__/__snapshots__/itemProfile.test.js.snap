// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemProfile /> renders with correct data 1`] = `
<h2>
  dogs are best
</h2>
`;

exports[`<ItemProfile /> renders with correct data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<ItemProfile /> renders with correct data 3`] = `
<p>
  dogs
</p>
`;
